File: GeneralsMD/Code/GameEngine/Source/GameClient/Statistics.cpp
// FILE: Statistics.cpp 
/*---------------------------------------------------------------------------*/
/* EA Pacific                                                                */
/* Confidential Information	                                                 */
/* Copyright (C) 2001 - All Rights Reserved                                  */
/* DO NOT DISTRIBUTE                                                         */
/*---------------------------------------------------------------------------*/
/* Project:    RTS3                                                          */
/* File name:  Statistics.cpp                                                */
/* Created:    John K. McDonald, Jr., 4/2/2002                               */
/* Desc:       Statistical functions should live here                        */
/* Revision History:                                                         */
/*		4/2/2002 : Initial creation                                            */
/*---------------------------------------------------------------------------*/

#include "PreRTS.h"	// This must go first in EVERY cpp file int the GameEngine

#include "GameClient/Statistics.h"

// Solution taken from http://www.epanorama.net/documents/telecom/ulaw_alaw.html
Real MuLaw(Real valueToRun, Real maxValueForVal, Real mu)
{
	Real testVal = (valueToRun - maxValueForVal / 2) / (maxValueForVal / 2);
	return (sign(testVal) * log(1 + mu * fabs(testVal)) / 
														 log(1 + mu));
}

// from my head. jkmcd
Real Normalize(Real valueToNormalize, Real minRange, Real maxRange)
{
	return ((valueToNormalize - minRange) / (maxRange - minRange));
}

// from my head again. jkmcd
Real NormalizeToRange(Real valueToNormalize, Real minRange, Real maxRange, Real outRangeMin, Real outRangeMax)
{
	return (Normalize(valueToNormalize, minRange, maxRange) * (outRangeMax - outRangeMin)) + outRangeMin;
}

File: GeneralsRust/Code/GameEngine/GameClient/src/statistics.rs
//! Statistical helper functions (ported from GameClient/Statistics.cpp).

/// Apply mu-law companding.
pub fn mu_law(value: f32, max_value: f32, mu: f32) -> f32 {
    let test_val = (value - max_value / 2.0) / (max_value / 2.0);
    test_val.signum() * (1.0 + mu * test_val.abs()).ln() / (1.0 + mu).ln()
}

/// Normalize a value from min/max to 0..1.
pub fn normalize(value: f32, min_range: f32, max_range: f32) -> f32 {
    (value - min_range) / (max_range - min_range)
}

/// Normalize to an arbitrary output range.
pub fn normalize_to_range(
    value: f32,
    min_range: f32,
    max_range: f32,
    out_min: f32,
    out_max: f32,
) -> f32 {
    normalize(value, min_range, max_range) * (out_max - out_min) + out_min
}
